RDoc-1763 (#1192)
ravendb/docs: Documentation/5.0/Raven.Documentation.Pages/client-api/session/configuration/use-session-context-for-load-balancing.markdown

# Use Session Context for Load Balancing

{NOTE: }

* The `LoadBalanceBehavior` convention determines whether a client session 
  is allowed to select the topology that would handle its requests.  
  When enabled, a session can select a topology by `tag`. Client requests 
  sent by sessions that use the same tag, are handled by the same topology.  
* Clients and administrators can use tags to load-balance traffic, e.g. by 
  using the topology tagged "users/1-A" for requests sent by this user.  

> 

Reviewer: ```suggestion

For example, consider a database with three nodes: `A`,`B` and `C`. With `LoadBalanceBehavior` set to `None`, the 
cluster will designate one of the nodes in the cluster as the primary, which will be used for both reads and writes.
You can use the `ReadBalanceBehavior` to spread the load among all the nodes of the database, but writes will
always go to the same primary node (which may change in case of failure).

he `LoadBalanceBehavior` convention allows you to specify a `tag`, which will mean that all sessions sharing
the same tag will have a _different_ primary node in the database topology. For example, if you are using a per
user context, sessions using the tag `users/1-A` will use node B as the primary (for reads and writes) while 
sessions using `users/2-A` will use node C as the primary.

Aside for the primary selection by the session, RavenDB continues to behave in the same manner. Replication 
between nodes in the database will operate normally and writes that were sent to one node would be replicate
to all the other nodes. 

This feature is useful when you have a degree of separation in your system, where a certain set of sessions 
will typically work on specific set of documents. Common examples include setting the context to the current user
or the current tenant. 
This will allow the client side to spread the write and read load across the nodes in the cluster on a fair basis

Failure of a particular node is handled normally, even if it is the primary node for a particular `tag` and will not
be visible to your code. Fail over to another node in the database will happen automatically and transparently. 

```

Author: reshaped the page
* there's a "LoadBalanceBehavior Usage" panel
* revised the information about administrator interferences in the "LoadBalanceBehavior Options" panel

{NOTE/}

{PANEL: LoadBalanceBehavior Options}

  * `None`  
    Read requests are handled based on the 
    [ReadBalanceBehavior convention](../../../client-api/configuration/load-balance-and-failover#readbalancebehavior-options).  
    Write requests are handled by the [Preferred Node](../../../client-api/configuration/load-balance-and-failover#preferred-node) 
    calculated by the client.  

  * `UseSessionContext`
     * A client session can select the topology that would handle its requests.  
       The topology is selected by `tag`.  
       Requests that use the same tag, are served by the same topology.  
       The same nodes are used for **Read** and **Write** requests.  
     * Administrators can disable this feature, or add a hash to randomize 
       the topology that would be selected for the client when it uses a tag.  
     * Using this option to choose client request-handling topology 
       influences only the client, causing no change in replication 
       or other server functions.  

  {NOTE: }
  Conflicts may rarely happen, when multiple sessions that use different tags 

>   modify a shared document concurrently. 

Reviewer: ```suggestion
  modify a shared document concurrently. This is handled by RavenDB according to your conflict resolution strategy.
 
```

And add a link to the relevant docs.

Author: done

# Use Session Context for Load Balancing

{NOTE: }

* The `LoadBalanceBehavior` convention determines whether the nodes 
  topology used to handle client requests, can be chosen by a client session.  
* When enabled, a session can select its handling topology by `tag`.  
  Client requests that use the same tag, are served by the same topology.  

> * Using pre-set topologies enables clients and administrators to load-balance requests 

Reviewer: ```suggestion
* Using preset topologies enables clients and administrators to load-balance requests 

```

Author: done

  as they please. E.g., frequent employee-related and product-related requests can be 

>   handled by an "**employees_top**" and a "**products_top**" topologies, respectively.  

Reviewer: I don't understand what you mean here

Author: is this clearer? - 
* Clients and administrators can use tags to load-balance traffic, e.g. by requesting **employee** documents from a topology tagged "employee" and **product** documents from a topology tagged "products".

Reviewer: That isn't how it works. 
You'll typically use something like the user id, or the tenant id to have the same topology for user / tenant.
That isn't on a per collection basis.

Author: i see. thanks.
can you give an example please, to a situation in which a user would want to always use the same topology?

* In this page:  
   * [LoadBalanceBehavior Options](../../../client-api/session/configuration/use-session-context-for-load-balancing#loadbalancebehavior-options)  

{NOTE/}

{PANEL: LoadBalanceBehavior Options}

  * `None`  
    Nodes are chosen to serve read requests by the 
    [ReadBalanceBehavior convention](../../../client-api/configuration/load-balance-and-failover#readbalancebehavior-options), 

>     and to server write requests by the [Preferred Node](../../../client-api/configuration/load-balance-and-failover#preferred-node) 

Reviewer: ```suggestion
    and to serve write requests by the [Preferred Node](../../../client-api/configuration/load-balance-and-failover#preferred-node) 

```

Reviewer: That said, the phrasing is awkward

Author: * `None`  
    Read requests are handled based on the [ReadBalanceBehavior convention](../../../client-api/configuration/load-balance-and-failover#readbalancebehavior-options).  
    Write requests are handled by the [Preferred Node](../../../client-api/configuration/load-balance-and-failover#preferred-node) calculated by the client.

Reviewer: ok

    calculated by the client.  

  * `UseSessionContext`
     * A client session can select the topology that would handle its requests.  
       The topology is selected by `tag`.  
       Requests that use the same tag, are served by the same topology.  
       The same nodes are used for **Read** and **Write** requests.  
     * Administrators can override client selections, selecting the 

>        topology that would handle client requests by its hash seed.  

Reviewer: Unclear what this means. Admins can either disable this feature or add a hash that would mix the topology if needed (rare)

Author: what does "mix the topology" mean?
for now - 

* Administrators can disable this feature or add a hash to choose the topology.

Reviewer: The hash they add is used to randomize the topology that is selected for the client when they set a tag

Author: done.

* Administrators can disable this feature, or add a hash to randomize the topology that would be selected for the client when it uses a tag.

     * Using this option to choose client request-handling topology 
       influences only the client, causing no change in replication 
       or other server functions.  

  {WARNING: }

>   Enabling this feature increases the chance for conflicts, 

Reviewer: Can we tone down the alarm level here?
This is usually not going to happen because in this model, you'll only write to nodes on your own tags

Author: {NOTE: }
  Conflicts may rarely happen, when multiple sessions that use different tags 
  modify a shared document concurrently. 
  {NOTE/}
